module/zfs/sa.c

```diff
@@ -1325,6 +1325,9 @@ sa_idx_tab_rele(objset_t *os, void *arg)
 	if (idx_tab == NULL)
 		return;
 
+	if (zfs_refcount_remove_unless_last(&idx_tab->sa_refcount, NULL))
+		return;
+
```

Comment:
Since `zfs_refcount_remove()` operates atomically adding a new `zfs_refcount_remove_unless_last()` functions seems redundant.  Can't we just move the the `mutex_enter/exit` call in to the `if (zfs_refcount_remove(&idx_tab->sa_refcount, NULL) == 0) { }` block below?  Or is there a case I'm overlooking?

Reply:
My routine retains the invariant that sa_refcount of 0 wont be spotted by anyone taking the lock below.

Even this happens to not be necessary to preserve for this particular case, there will be others.

Reply:
Fair enough.